src/MSIdentityScaffolding/Microsoft.DotNet.MSIdentity/CodeReaderWriter/ProjectModifier.cs

```diff
@@ -40,62 +40,130 @@ public ProjectModifier(ApplicationParameters applicationParameters, Provisioning
+            }
+        }
+
+        private async Task HandleCodeFileAsync(CodeFile file, CodeAnalysis.Project project, CodeChangeOptions options)
+        {
+            if (string.IsNullOrEmpty(file.FileName))
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(file.AddFilePath))
+            {
+                AddFile(file);
+                return;
+            }
+
+            var fileName = file.FileName.Equals("Startup.cs", StringComparison.OrdinalIgnoreCase)
+                    ? await ProjectModifierHelper.GetStartupClass(_toolOptions.ProjectPath, project)
+                    : file.FileName;
+
+            if (fileName.Equals("Program.cs"))
+            {
+                // only modify Program.cs file if it's a minimal hosting app (as we made changes to the Startup.cs file).
+                if (options.IsMinimalApp)
+                {
+                    await ModifyProgramCs(file, project, options);
                 }
             }
+            else if (fileName.EndsWith(".cs"))
+            {
+                await ModifyCsFile(fileName, file, project, options);
+            }
+            else if (fileName.EndsWith(".cshtml"))
+            {
+                await ModifyCshtmlFile(fileName, file, project, options);
+            }
+            else if (fileName.EndsWith(".razor"))
+            {
+                await ModifyRazorFile(fileName, file, project, options);
+            }
+        }
+
+        private void AddFile(CodeFile file)
+        {
+            var resourceName = file.FileName.Replace('.', '_');
+            var propertyInfo = AppProvisioningTool.Properties.Where(
+                p => p.Name.StartsWith("add") && p.Name.EndsWith(resourceName)).FirstOrDefault();
+
+            if (propertyInfo is null)
+            {
+                return;
+            }
+
+            byte[] content = (propertyInfo.GetValue(null) as byte[])!;
+
+            string codeFileString = Encoding.UTF8.GetString(content);
+
+            if (string.IsNullOrEmpty(codeFileString))
+            {
+                throw new FormatException($"Resource file { propertyInfo.Name } could not be parsed. ");
+            }
+
+            var filePath = Path.Combine(_toolOptions.ProjectPath, file.AddFilePath);
+            if (File.Exists(filePath))
+            {
+                return;
+            }
+
+            var fileDir = Path.GetDirectoryName(filePath);
+
+            if (!string.IsNullOrEmpty(fileDir))
+            {
+                Directory.CreateDirectory(fileDir);
+                File.WriteAllText(filePath, codeFileString);
+            }
+        }
+
+        internal async Task ModifyRazorFile(string fileName, CodeFile file, CodeAnalysis.Project project, CodeChangeOptions toolOptions)
+        {
+            string? filePath = Directory.EnumerateFiles(_toolOptions.ProjectPath, fileName, SearchOption.AllDirectories).FirstOrDefault();
```

Comment:
This file path should be discoverable from the file paths in CodeAnalysis.Project.
Directory.EnumerateFiles is a much more expensive operation, especially for large projects.

Reply:
You do it down below, just have that check, just check using `project.Documents.Where(d => d.Name.EndsWith.Equals(yaddayadda)`